src/mon/PGMonitor.cc

```diff
@@ -1234,6 +1235,146 @@ bool PGMonitor::preprocess_command(MonOpRequestRef op)
                           mon->osdmon()->osdmap.get_inst(osd));
     ss << "instructing pg " << pgid << " on osd." << osd << " to " << scrubop;
     r = 0;
+  } else if (prefix == "pg force-recovery" ||
+             prefix == "pg force-backfill" ||
+             prefix == "pg cancel-force-recovery" ||
+             prefix == "pg cancel-force-backfill") {
+    string forceop = prefix.substr(3, string::npos);
+    list<pg_t> parsed_pgs;
+
+    // figure out actual op just once
+    int actual_op = 0;
+    if (forceop == "force-recovery") {
+      actual_op = OFR_RECOVERY;
+    } else if (forceop == "force-backfill") {
+      actual_op = OFR_BACKFILL;
+    } else if (forceop == "cancel-force-backfill") {
+      actual_op = OFR_BACKFILL | OFR_CANCEL;
+    } else if (forceop == "cancel-force-recovery") {
+      actual_op = OFR_RECOVERY | OFR_CANCEL;
+    }
+
+    // covnert pg names to pgs, discard any invalid ones while at it
+    {
+      // we don't want to keep pgidstr and pgidstr_nodup forever
+      vector<string> pgidstr;
+      set<string> pgidstr_nodup;
+
+      // get pgids to process and prune duplicates
+      cmd_getval(g_ceph_context, cmdmap, "pgid", pgidstr);
+      for (auto& pstr : pgidstr) {
```

Comment:
you should be able to simply do:

    set<string> pgidstr_nodup(pgidstr.begin(), pgidstr.end())